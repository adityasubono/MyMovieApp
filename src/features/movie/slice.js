import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  movies: [],
  totalResults: 0,
  success: false,
  movie: {}
};

const movieSlice = createSlice({
    name: 'movie',
    initialState,
    reducers: {
      getListSuccess: (state, action) => {
        state.success = true;
        state.movies = (action.payload.movies && 
          action.payload.movies.map((item) => ({ 
            ...item, Poster: item.Poster !== 'N/A' ? item.Poster : null
          }))) || [];

        state.totalResults = action.payload.totalResults;
      },
      getListError: (state, action) => {
        state.movies = [];
        state.totalResults = 0;
        state.success = false;
      },
      detailSuccess: (state, action) => {
        const movie = action.payload.movie;

        state.success = true;
        state.movie = movie;
        state.movie.Actors = movie.Actors.split(', ') || [];
        state.movie.Genre = movie.Genre.split(', ') || [];
        state.movie.Poster = movie.Poster !== 'N/A' ? movie.Poster : null;
      },
      detailLoading: (state) => {
        state.success = false;
        state.movie = {};
      }
    },
  });

export const { getListSuccess, getListError, detailSuccess, detailLoading } = movieSlice.actions;

export default movieSlice.reducer;
