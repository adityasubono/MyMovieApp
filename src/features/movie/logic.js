import { createAsyncThunk } from '@reduxjs/toolkit';
import axiosInstance from '../../app/axiosConfig';

import { getListSuccess, detailSuccess, getListError, detailLoading } from './slice';

export const getList = createAsyncThunk(
    'movie/fetchListMovie',
    async ({ searchTerm, page }, thunkAPI) => {
      const response = await axiosInstance(`/?s=${searchTerm}&page=${page}`);

      if (response.data.Response === "False") {
        thunkAPI.dispatch(getListError());
        return response.data;
      }
      
      thunkAPI.dispatch(getListSuccess({ movies: response.data.Search, totalResults: (response.data.totalResults && parseInt(response.data.totalResults)) || 0 }));
      return response.data;
    }
  );


export const detailMovieById = createAsyncThunk(
  'movie/fetchDetailMovie',
  async (id, thunkAPI) => {
    thunkAPI.dispatch(detailLoading());
    const response = await axiosInstance(`/?i=${id}`);
    
    thunkAPI.dispatch(detailSuccess({ movie: response.data }));
    return response.data;
  }
);