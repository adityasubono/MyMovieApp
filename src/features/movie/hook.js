import { useEffect, useState, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {getList, detailMovieById} from './logic';


export const useMovies = () => {
    const dispatch = useDispatch();
    const moviesRaw = useSelector((state) => state.movie.movies);
    const isSuccess = useSelector((state) => state.movie.success);

    const [searchTerm, setSearchTerm] = useState('');
    const [page, setPage] = useState(1);
    const [loading, setLoading] = useState(false);
    const [movies, setMovies] = useState([]);

    useEffect(() => {
        /**
        * This function will fetch data from the API
        * with given parameter from the modules
        */
        setLoading(true);
        const dispatcher = dispatch(getList({ searchTerm, page }));

        dispatcher.then(() => {
            setLoading(false);
        });

        return () => {
            dispatcher.abort();
            setLoading(false);
        };
    }, [searchTerm]);

    useEffect(() => {
        setMovies([]);
    }, [searchTerm]);

    useEffect(() => {
        /**
         * This function will create a new array with existing list data and
         * additional data from the API based on pagination and wil return array with unique item
         */
         if (!isSuccess) {
            setMovies([]);
         } else {
             const key = 'imdbID';
             setMovies((prev) => {
                 return  [...new Map([...prev, ...moviesRaw].map(item =>
                     [item[key], item])).values()];
               });
         }
    }, [page, searchTerm, moviesRaw]);

    const onLoadMore = useCallback(async () => {
        setLoading(true);
        await dispatch(getList({ searchTerm, page: page + 1 }));
        setLoading(false);
        setPage(page + 1);
    }, [page, searchTerm]);

    return {loading, movies, searchTerm, setSearchTerm, onLoadMore, setPage};
}

export const useMovie = (id) => {
    const dispatch = useDispatch();
    const movie = useSelector((state) => state.movie.movie);

    const [loading, setLoading] = useState(false);

    useEffect(() => {
        /**
        * This function will fetch data from the API
        * with given parameter from the modules
        */
        setLoading(true);
        const dispatcher = dispatch(detailMovieById(id));

        dispatcher.then(() => {
            setLoading(false);
        });

        return () => {
            dispatcher.abort();
            setLoading(false);
        };
    }, []);

    return {movie, loading}
}
