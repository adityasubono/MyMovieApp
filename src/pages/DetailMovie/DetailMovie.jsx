import React, { useCallback, useState } from 'react';

import { useMovie } from '../../features/movie/hook';

import DefaultPoster from '../../assets/default_poster.png'

const DetailMovie = ({ match }) => {
    const [activeTab, setActiveTab] = useState('details');

    const {movie} = useMovie(match.params.id);

    const onChangeTab = useCallback((id) => () => {
        setActiveTab(id);
    }, []);

    if (!movie) {
        return (null);
    }

    return (
        <div className="row mt-5 justify-content-center h-100">
            <div className="col-md-2 p-3 ">
                <div className="row justify-content-center">
                    <img className="poster"  src={movie.Poster || DefaultPoster} alt="" data-testid="default-img"/>
                </div>
            </div>
            <div className="col-md-8 p-3">
                <div className="row">
                    <div className="col-md-10" style={{ paddingLeft: '20px' }}>
                        <h2>{movie.Title}</h2>
                        <span>{movie.Year}</span>
                        <span style={{ marginLeft: '10px' }}>Directed by <b>{movie.Director}</b></span>
                        <span className="badge bg-primary" style={{ marginLeft: '15px' }}>{movie.Rated}</span>
                    </div>
                </div>
                <div className="row">
                    <div className="col-md-8">
                        <div className="pt-3 w-75">
                            <p style={{ paddingLeft: '10px' }}>{movie.Plot}</p>
                        </div>

                        <ul className="nav nav-tabs mt-4" data-testid="nav-tabs">
                            <li className="nav-item">
                                <button onClick={onChangeTab('details')}
                                    className={activeTab === 'details' ? "nav-link active" : "nav-link"}>
                                        Details
                                    </button>
                            </li>
                            <li className="nav-item">
                                <button onClick={onChangeTab('cast')}
                                    className={activeTab === 'cast' ? "nav-link active" : "nav-link"}>
                                        Cast
                                    </button>
                            </li>
                            <li className="nav-item">
                                <button onClick={onChangeTab('genre')}
                                    className={activeTab === 'genre' ? "nav-link active" : "nav-link"}>
                                        Genre
                                    </button>
                            </li>
                        </ul>
                        <div className="row p-4">
                            {activeTab === 'cast' && <div className="col-md-12">
                                {movie.Actors && movie.Actors.map((actor, i) => <span class="badge bg-secondary m-1" key={i}>{actor}</span>)}
                            </div>}
                            {activeTab === 'details' && <div className="col-md-12">
                                <b>Languange: </b><p>{movie.Language}</p>
                                <b>Country: </b><p>{movie.Country}</p>
                                <b>Awards: </b><p>{movie.Awards}</p>
                                <b>Production: </b><p>{movie.Production}</p>
                                <b>BoxOffice: </b><p>{movie.BoxOffice}</p>
                            </div>}
                            {activeTab === 'genre' && <div className="col-md-12">
                                {movie.Genre && movie.Genre.map((gen, i) => <span class="badge bg-secondary m-1" key={i}>{gen}</span>)}
                            </div>}
                        </div>
                    </div>
                    <div className="col-md-4 d-flex">
                        {movie.Ratings && movie.Ratings.map((item, i) => (<div className="m-3 mt-0" key={i}>
                            <div className="d-flex align-items-end rating-title"><p>{item.Source}</p></div>
                            <span className="badge bg-primary">{item.Value}</span>
                        </div>))}
                    </div>
                </div>
            </div>
        </div>
    );
};

export default DetailMovie;