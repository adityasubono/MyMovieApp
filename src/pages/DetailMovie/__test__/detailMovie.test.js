import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { shallow } from 'enzyme';

import  * as hooks from '../../../features/movie/hook';

import DetailMovie from '../DetailMovie';

import DefaultPoster from '../../../assets/default_poster.png';

const dummyData = {
    "Title": "Batman",
    "Year": "1989",
    "Rated": "PG-13",
    "Released": "23 Jun 1989",
    "Runtime": "126 min",
    "Genre": "Action, Adventure",
    "Director": "Tim Burton",
    "Writer": "Bob Kane, Sam Hamm, Warren Skaaren",
    "Actors": "Michael Keaton, Jack Nicholson, Kim Basinger",
    "Plot": "Gotham City. Crime boss Carl Grissom (Jack Palance) effectively runs the town but there's a new crime fighter in town - Batman (Michael Keaton). Grissom's right-hand man is Jack Napier (Jack Nicholson), a brutal man who is not entirely sane... After falling out between the two Grissom has Napier set up with the Police and Napier falls to his apparent death in a vat of chemicals. However, he soon reappears as The Joker and starts a reign of terror in Gotham City. Meanwhile, reporter Vicki Vale (Kim Basinger) is in the city to do an article on Batman. She soon starts a relationship with Batman's everyday persona, billionaire Bruce Wayne.",
    "Language": "English, French, Spanish",
    "Country": "United States, United Kingdom",
    "Awards": "Won 1 Oscar. 9 wins & 26 nominations total",
    "Poster": "https://m.media-amazon.com/images/M/MV5BMTYwNjAyODIyMF5BMl5BanBnXkFtZTYwNDMwMDk2._V1_SX300.jpg",
    "Ratings": [
      {
        "Source": "Internet Movie Database",
        "Value": "7.5/10"
      },
      {
        "Source": "Rotten Tomatoes",
        "Value": "71%"
      },
      {
        "Source": "Metacritic",
        "Value": "69/100"
      }
    ],
    "Metascore": "69",
    "imdbRating": "7.5",
    "imdbVotes": "345,938",
    "imdbID": "tt0096895",
    "Type": "movie",
    "DVD": "24 Jul 2014",
    "BoxOffice": "$251,348,343",
    "Production": "Warner Brothers, PolyGram Filmed Entertainment, Guber-Peters Company",
    "Website": "N/A",
    "Response": "True"
  };

describe('Detail Movie Component Test', () => {
  let routeComponentPropsMock;
    afterEach(cleanup)
    beforeEach(() => {
      routeComponentPropsMock = {
        history: {},
        location: {},
        match: {
            params: {
                id: 1
            }
        },
      };
      jest.spyOn(hooks, 'useMovie').mockImplementation(() => ({movie: dummyData}));
    })
    it('should render detail movie component and match with snapshot', () => {
        const wrapper = shallow(<DetailMovie {...routeComponentPropsMock}/>);
        expect(wrapper.getElements()).toMatchSnapshot();
    });

    it('should render image poster with default image', () => {
      dummyData.Poster = null;
      const wrapper = render(<DetailMovie {...routeComponentPropsMock}/>);
      const result = wrapper.getByTestId('default-img');
      expect(result).toHaveAttribute('src', DefaultPoster)
    });

    // eslint-disable-next-line jest/no-identical-title
    it('should render image poster with default image', () => {
      const wrapper = render(<DetailMovie {...routeComponentPropsMock}/>);
      const result = wrapper.getByTestId('nav-tabs')
      expect(result).toBeTruthy();
      expect(result).toBeInTheDocument();
    });
})
