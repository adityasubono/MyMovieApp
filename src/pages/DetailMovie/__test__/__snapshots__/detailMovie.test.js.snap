// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Movie Component Test should render detail movie component and match with snapshot 1`] = `
Array [
  <div
    className="row mt-5 justify-content-center h-100"
  >
    <div
      className="col-md-2 p-3 "
    >
      <div
        className="row justify-content-center"
      >
        <img
          alt=""
          className="poster"
          data-testid="default-img"
          src="https://m.media-amazon.com/images/M/MV5BMTYwNjAyODIyMF5BMl5BanBnXkFtZTYwNDMwMDk2._V1_SX300.jpg"
        />
      </div>
    </div>
    <div
      className="col-md-8 p-3"
    >
      <div
        className="row"
      >
        <div
          className="col-md-10"
          style={
            Object {
              "paddingLeft": "20px",
            }
          }
        >
          <h2>
            Batman
          </h2>
          <span>
            1989
          </span>
          <span
            style={
              Object {
                "marginLeft": "10px",
              }
            }
          >
            Directed by 
            <b>
              Tim Burton
            </b>
          </span>
          <span
            className="badge bg-primary"
            style={
              Object {
                "marginLeft": "15px",
              }
            }
          >
            PG-13
          </span>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-8"
        >
          <div
            className="pt-3 w-75"
          >
            <p
              style={
                Object {
                  "paddingLeft": "10px",
                }
              }
            >
              Gotham City. Crime boss Carl Grissom (Jack Palance) effectively runs the town but there's a new crime fighter in town - Batman (Michael Keaton). Grissom's right-hand man is Jack Napier (Jack Nicholson), a brutal man who is not entirely sane... After falling out between the two Grissom has Napier set up with the Police and Napier falls to his apparent death in a vat of chemicals. However, he soon reappears as The Joker and starts a reign of terror in Gotham City. Meanwhile, reporter Vicki Vale (Kim Basinger) is in the city to do an article on Batman. She soon starts a relationship with Batman's everyday persona, billionaire Bruce Wayne.
            </p>
          </div>
          <ul
            className="nav nav-tabs mt-4"
            data-testid="nav-tabs"
          >
            <li
              className="nav-item"
            >
              <button
                className="nav-link active"
                onClick={[Function]}
              >
                Details
              </button>
            </li>
            <li
              className="nav-item"
            >
              <button
                className="nav-link"
                onClick={[Function]}
              >
                Cast
              </button>
            </li>
            <li
              className="nav-item"
            >
              <button
                className="nav-link"
                onClick={[Function]}
              >
                Genre
              </button>
            </li>
          </ul>
          <div
            className="row p-4"
          >
            <div
              className="col-md-12"
            >
              <b>
                Languange: 
              </b>
              <p>
                English, French, Spanish
              </p>
              <b>
                Country: 
              </b>
              <p>
                United States, United Kingdom
              </p>
              <b>
                Awards: 
              </b>
              <p>
                Won 1 Oscar. 9 wins & 26 nominations total
              </p>
              <b>
                Production: 
              </b>
              <p>
                Warner Brothers, PolyGram Filmed Entertainment, Guber-Peters Company
              </p>
              <b>
                BoxOffice: 
              </b>
              <p>
                $251,348,343
              </p>
            </div>
          </div>
        </div>
        <div
          className="col-md-4 d-flex"
        >
          <div
            className="m-3 mt-0"
          >
            <div
              className="d-flex align-items-end rating-title"
            >
              <p>
                Internet Movie Database
              </p>
            </div>
            <span
              className="badge bg-primary"
            >
              7.5/10
            </span>
          </div>
          <div
            className="m-3 mt-0"
          >
            <div
              className="d-flex align-items-end rating-title"
            >
              <p>
                Rotten Tomatoes
              </p>
            </div>
            <span
              className="badge bg-primary"
            >
              71%
            </span>
          </div>
          <div
            className="m-3 mt-0"
          >
            <div
              className="d-flex align-items-end rating-title"
            >
              <p>
                Metacritic
              </p>
            </div>
            <span
              className="badge bg-primary"
            >
              69/100
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
